operator/controllers/btpoperator_controller.go

```diff
@@ -347,3 +417,283 @@ func (r *BtpOperatorReconciler) watchBtpOperatorUpdatePredicate() predicate.Func
+		}
+		if !resourcesLeft {
+			success <- true
+			return
+		}
+		time.Sleep(retryInterval)
+	}
+}
+
+func (r *BtpOperatorReconciler) hardDelete(ctx context.Context, gvk schema.GroupVersionKind, namespaces *corev1.NamespaceList) error {
+	logger := log.FromContext(ctx)
+	object := &unstructured.Unstructured{}
+	object.SetGroupVersionKind(gvk)
+
+	for _, namespace := range namespaces.Items {
+		if err := r.DeleteAllOf(ctx, object, client.InNamespace(namespace.Name)); err != nil {
+			logger.Error(err, "Err while doing delete all of")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) checkIfAnyResourcesLeft(ctx context.Context, namespaces *corev1.NamespaceList) (error, bool) {
+	anyLeft := func(namespace string, gvk schema.GroupVersionKind) (error, bool) {
+		list := &unstructured.UnstructuredList{}
+		list.SetGroupVersionKind(gvk)
+		if err := r.List(ctx, list, client.InNamespace(namespace)); err != nil {
+			if !errors.IsNotFound(err) {
+				return err, true
+			}
+		}
+
+		return nil, len(list.Items) > 0
+	}
+
+	for _, namespace := range namespaces.Items {
+		err, instancesLeft := anyLeft(namespace.Name, instanceGvk)
+		if err != nil {
+			return err, true
+		}
+		err, bindingsLeft := anyLeft(namespace.Name, bindingGvk)
+		if err != nil {
+			return err, true
+		}
+		if instancesLeft == false && bindingsLeft == false {
+			return nil, false
+		}
+	}
+
+	return nil, true
+}
+
+func (r *BtpOperatorReconciler) handleSoftDelete(ctx context.Context, namespaces *corev1.NamespaceList) error {
+	logger := log.FromContext(ctx)
+	logger.Info("hard delete failed. trying to perform soft delete")
+
+	if err := r.softDelete(ctx, &instanceGvk); err != nil {
+		logger.Error(err, "soft deletion of instances failed")
+		return err
+	}
+
+	if err := r.softDelete(ctx, &bindingGvk); err != nil {
+		logger.Error(err, "hard deletion of bindings failed")
+		return err
+	}
+
+	if err := r.cleanUpAllBtpOperatorResources(ctx, namespaces); err != nil {
+		logger.Error(err, "failed to remove related installed resources")
+		return err
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) softDelete(ctx context.Context, gvk *schema.GroupVersionKind) error {
+	listGvk := *gvk
+	listGvk.Kind = gvk.Kind + "List"
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(listGvk)
+
+	if err := r.List(ctx, list); err != nil {
+		err = fmt.Errorf("%w; could not list in soft delete", err)
+		return err
+	}
+
+	isBinding := gvk.Kind == btpOperatorServiceBinding
+	for _, item := range list.Items {
+		item.SetFinalizers([]string{})
+		if err := r.Update(ctx, &item); err != nil {
+			return err
+		}
+
+		if isBinding {
+			secret := &corev1.Secret{}
+			secret.Name = item.GetName()
+			secret.Namespace = item.GetNamespace()
+			if err := r.Delete(ctx, secret); err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := r.List(ctx, list); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		if len(list.Items) > 0 {
```

Comment:
won't this condition be always true where there are service instances/service bindings in the cluster? Before this condition, only Secrets are deleted and Finalizers are removed from resources.

Reply:
what about the case when DeleteAllOf() fails and `anyErr` is set to true?

Reply:
I think we may run into problems when deletion timestamp is not set and we reach softDelete().